rpc/rest/security/rest_security.go

```diff
@@ -0,0 +1,422 @@
+// Validate the request
+func (a *authenticator) Validate(provider http.HandlerFunc) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		// Token may be accessed via cookie, or from authentication header
+		var tokenString string
+		// Try to read the cookie first
+		cookie, err := req.Cookie(cookieName)
+		if err == nil && cookie != nil {
+			tokenString = cookie.Value
+		} else {
+			// Do not return error yet
+			a.log.Debugf("Authentication cookie not found (err: %v)", err)
+			// Continue with reading header
+			authHeader := req.Header.Get(authLabel)
+			if authHeader == "" {
+				a.formatter.Text(w, http.StatusUnauthorized, "401 Unauthorized: authorization required")
+				return
+			}
+			bearerToken := strings.Split(authHeader, " ")
+			if len(bearerToken) != 2 {
+				a.formatter.Text(w, http.StatusUnauthorized, "401 Unauthorized: invalid authorization token")
+				return
+			}
+			// Parse token header constant
+			if bearerToken[0] != "Bearer" {
+				a.formatter.Text(w, http.StatusUnauthorized, "401 Unauthorized: invalid authorization header")
+				return
+			}
+			tokenString = bearerToken[1]
+		}
+		// Retrieve token object from raw string
+		token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
+			if _, ok := jwt.GetSigningMethod(token.Header["alg"].(string)).(*jwt.SigningMethodHMAC); !ok {
+				return nil, fmt.Errorf("error parsing token")
+			}
+			return []byte(signature), nil
+		})
+		if err != nil {
+			errStr := fmt.Sprintf("500 internal server error: %s", err)
+			a.formatter.Text(w, http.StatusInternalServerError, errStr)
+			return
+		}
+		// Validate token claims
+		if token.Claims != nil {
+			if err := token.Claims.Valid(); err != nil {
+				errStr := fmt.Sprintf("401 Unauthorized: %v", err)
+				a.formatter.Text(w, http.StatusUnauthorized, errStr)
+				return
+			}
+		}
+		// Validate token itself
+		if err := a.validateToken(token, req.URL.Path, req.Method); err != nil {
+			errStr := fmt.Sprintf("401 Unauthorized: %v", err)
+			a.formatter.Text(w, http.StatusUnauthorized, errStr)
+			return
+		}
+
+		provider.ServeHTTP(w, req)
+	})
+}
+
+// Register authenticator-wide security handlers
+func (a *authenticator) registerSecurityHandlers() {
+	a.router.HandleFunc(login, a.loginHandler).Methods(http.MethodGet, http.MethodPost)
+	a.router.HandleFunc(authenticate, a.authenticationHandler).Methods(http.MethodPost)
+	a.router.HandleFunc(logout, a.logoutHandler).Methods(http.MethodPost)
+}
+
+// Login handler shows simple page to log in
+func (a *authenticator) loginHandler(w http.ResponseWriter, req *http.Request) {
+	// GET returns login page. Submit redirects to authenticate.
+	if req.Method == http.MethodGet {
+		r := render.New(render.Options{
+			Directory:  "templates",
+			Asset:      Asset,
+			AssetNames: AssetNames,
+		})
+		r.HTML(w, http.StatusOK, "login", nil)
+	}
+	// POST decodes provided credentials
+	if req.Method == http.MethodPost {
```

Comment:
Why is this not in else case? It needs to be evaluated only if first one is false.

Reply:
ok but honestly, do we really save something here?